lib/src/tls.rs

```diff
@@ -0,0 +1,1036 @@
+    ) -> Result<ParsedCertificateAndKey, Self::Error>;
+}
+
+// -----------------------------------------------------------------------------
+// CertificateOverride struct
+
+#[derive(Clone, Debug)]
+pub struct CertificateOverride {
+    pub names: Option<HashSet<String>>,
+    pub expiration: Option<i64>,
+}
+
+impl From<&AddCertificate> for CertificateOverride {
+    fn from(opts: &AddCertificate) -> Self {
+        let mut names = None;
+        if !opts.names.is_empty() {
+            names = Some(opts.names.iter().cloned().collect())
+        }
+
+        Self {
+            names,
+            expiration: opts.expired_at.to_owned(),
+        }
+    }
+}
+
+// -----------------------------------------------------------------------------
+// ParsedCertificateAndKey struct
+
+#[derive(Debug)]
+pub struct ParsedCertificateAndKey {
+    pub certificate: Pem,
+    pub chain: Vec<Pem>,
+    pub key: String,
+    pub versions: Vec<TlsVersion>,
+}
+
+impl Clone for ParsedCertificateAndKey {
+    fn clone(&self) -> Self {
+        let certificate = Pem {
+            label: self.certificate.label.to_owned(),
+            contents: self.certificate.contents.to_owned(),
+        };
+
+        Self {
+            certificate,
+            chain: self
+                .chain
+                .iter()
+                .map(|chain| Pem {
+                    label: chain.label.to_owned(),
+                    contents: chain.contents.to_owned(),
+                })
+                .collect(),
+            key: self.key.to_owned(),
+            versions: self.versions.to_owned(),
+        }
+    }
+}
+
+// -----------------------------------------------------------------------------
+// GenericCertificateResolverError enum
+
+#[derive(thiserror::Error, Clone, Debug)]
+pub enum GenericCertificateResolverError {
+    #[error("failed to parse certificate common name, {0}")]
+    CommonNameParseError(String),
+    #[error("failed to parse pem certificate, {0}")]
+    PemParseError(String),
+    #[error("failed to parse der certificate, {0}")]
+    DerParseError(String),
+    #[error("certificate, chain or private key is invalid")]
+    InvalidPrivateKeyError,
+    #[error("certificate is still in use")]
+    IsStillInUseError,
+}
+
+// -----------------------------------------------------------------------------
+// GenericCertificateResolver struct
+
+#[derive(Debug)]
+pub struct GenericCertificateResolver {
+    pub domains: TrieNode<CertificateFingerprint>,
+    certificates: HashMap<CertificateFingerprint, ParsedCertificateAndKey>,
+    name_fingerprint_idx: HashMap<String, HashSet<CertificateFingerprint>>,
+    overrides: HashMap<CertificateFingerprint, CertificateOverride>,
+}
+
+impl CertificateResolver for GenericCertificateResolver {
+    type Error = GenericCertificateResolverError;
+
+    fn get_certificate(
+        &self,
+        fingerprint: &CertificateFingerprint,
+    ) -> Option<ParsedCertificateAndKey> {
```

Comment:
this should return a Result to allow Listener implementations of CertificateResolver to return an error (and this would fix the CI failure at lib/src/https_openssl.rs#L1141 )

Reply:
It was a misused of `map_err` method in the openssl context. It has to be an `Option` as the meaning is to indicate the existence of the object and not a way to handle an error.